### DISCUSSION TITLE
How do I get a list of userids from players in the match

### DISCUSSION THREAD

**Discussion by KyeNormanGill** [saul/demofile#207] [Questions] [answered]
I've looked through the docs but i can't seem to find how to get userids so i can match them with names and steamids etc.

**Comment by saul** [comment, marked as answer]
```javascript
demo.players.map(p => p.userId) // will return all player userids
demo.entities.getByUserId(userid) // will return a player that has a specifc userid
```

**Comment by KyeNormanGill** [comment]
The players map returns `[ undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined ]`

**Comment by KyeNormanGill** [comment]
Oh nevermind. Capitol I on userId.

**Comment by saul** [comment]
Thanks - I've updated my comment for future reference

**Comment by dnnrwttr** [comment]
@saul what about bots in a demo file? If i read in competitive matches and fight versus a bot, cause one opponent was kicked?

i have a concrete problem:
i have a demo where a player was kicked and after the kick a new userid is in the events... in my case the 22.
the problem ist no player has the 22... i checked the player info and there the bot has the 10 (like the REAL player before kick/leave). what can i do now?

**Comment by saul** [comment]
That's bizarre - I was under the impression that bots that come in to replace players have a new user ID.

What problems is this causing you?

**Comment by dnnrwttr** [comment]
@saul maybe you download the following demofile: https://www.mediafire.com/?ov4arcjvbc2hd67 or see my attached json file...

i put it in your matchparser and only "echo" the player_death and round_prestart event.
In the 19. (first kill) and 20. (3rd kill) round. you will see the following constellation:

`player_death {"userid":22,"attacker":5,....`

the userid 22 isn't known... if i echo the player infos i get: no userid with 22...
Player 22 is "Bot Niles" and have the ID 10 in the player infos....

i think that is the problem in my case...

here the output:
https://img.esnt.de/6pQ5i

can you help me?

----

*edit*

i get problems if i try to get 
`var victim = demo.entities.getByUserId(e.userid);`

**Comment by dnnrwttr** [comment]
here is the proof of it... i found the following:

player_connect {"name":"Niles","index":10,"userid":22,"networkid":"BOT","address":""}
player_team {"userid":22,"team":2,"oldteam":0,"disconnect":false,"autoteam":false,"silent":false,"isbot":true}

that should mean that BOT Niles got the UserID 22 from UserID10 or? But why does the getByUserId dont realize this info?

**Comment by saul** [comment]
No it means a bot has joined as userid 22. The index is the entity index.

**Comment by dnnrwttr** [comment]
hm did you understand my prob. saul?

**Comment by saul** [comment]
I'm not entirely sure - are you saying that when the bot joins, you can't get any info about the player at user ID 22?

**Comment by dnnrwttr** [comment]
thats correct.

but i dont know if that is my fault or its your part... maybe you download my demo file (link above) and have a look at the given rounds.

**Comment by dnnrwttr** [comment]
@saul i just want to try to clearify that:

i save the kills from every player in a database... if the bot is connecting (and the real player disconnecting) i dont get the kills!

so if i killed the bot - the kill will not be saved, cause i THINK that i dont get the infos with the 
`var victim = demo.entities.getByUserId(e.userid);` u know?

cause the bot have an other ID as the player before

**Comment by saul** [comment]
I've just tried the demo for myself - `getByUserId(22)` works fine for me - I can get info for Niles no problem. I'm not sure what your bug is? New players (e.g. a bot that takes over a disconnecting player) will be given a new user ID.
